src/integer.h

```diff
@@ -55,16 +55,26 @@ GIT_INLINE(bool) git__add_uint64_overflow(uint64_t *out, uint64_t one, uint64_t
 }
 
 /* Use clang/gcc compiler intrinsics whenever possible */
-#if (SIZE_MAX == ULONG_MAX) && __has_builtin(__builtin_uaddl_overflow)
-# define git__add_sizet_overflow(out, one, two) \
-	__builtin_uaddl_overflow(one, two, out)
-# define git__multiply_sizet_overflow(out, one, two) \
-	__builtin_umull_overflow(one, two, out)
-#elif (SIZE_MAX == UINT_MAX) && __has_builtin(__builtin_uadd_overflow)
-# define git__add_sizet_overflow(out, one, two) \
-	__builtin_uadd_overflow(one, two, out)
-# define git__multiply_sizet_overflow(out, one, two) \
-	__builtin_umul_overflow(one, two, out)
+#if (__has_builtin(__builtin_add_overflow) || \
+     (defined(__GNUC__) && (__GNUC__ >= 5)))
+
+#	if (ULONG_MAX == ULLONG_MAX) && defined(_WIN64)
+#		define git__add_sizet_overflow(out, one, two) \
+			__builtin_uaddll_overflow(one, two, out)
+#		define git__multiply_sizet_overflow(out, one, two) \
+			__builtin_umulll_overflow(one, two, out)
+#	elif (ULONG_MAX == ULLONG_MAX)
```

Comment:
not by definition, since `unsigned long` and `unsigned long long` are still different types to the compiler, even though they have the same size. so passing in a `size_t` to the wrong method results in a warning, which goes against what i'm trying to achieve.

Reply:
But you're not doing anything here to validate that `size_t` is actually _either_ of those types.  Again, not likely but theoretically possible.

Reply:
In other words, couldn't we end up in a (yes, I admit, almost certainly theoretical) situation where `unsigned long` and `unsigned long long` are both 64 bits but `size_t` is something shorter?

Thanks for explaining the rationale behind the `uaddll` vs `uaddl`, that makes more sense now.  Some comments would indeed be helpful because this is super not obvious to me.

Reply:
err... i totally missed that case. will add that explicit check together with the comments.

Reply:
I think it's exceeeeeeeedingly unlikely to ever occur in practice.  If you wanted to dust off your Cray and port libgit2 to it then it would still not have a modern compiler or 64 bit `unsigned long`s.  But I like being a little extra paranoid when it comes to this little bit of security.

Thanks for accommodating my paranoia.